### PR TITLE
Debounce and fix for default values in serialization filter 

### PR DESCRIPTION
**Pull request by TFTomSun**
I've added the possibility to debounce callbacks. That reduces the load on browser and network level for callbacks that are invoked very often.

I also found a bug in the serialization filter behavior. I used something like new { myValue = default(int) }. myValue was not included into the serialization because !currentMemberSpec evaluates to false. On the other hand that approach of filtering allowed me to directly use that anonymous type as a deserialization target  type. Specifying something like new {myValue = 1} seems not to be intuitive. 
To avoid breaking changes I made the default value handling configurable.

### CODE REVIEW

**Comment by RemiBou** [comment]
Wow, sorry didn't receive an email for this PR, I'll review it ASAP. Thanks a lot I'm honored you spend some time on this !

**Comment by TFTomSun** [comment]
You're welcome. I have to thank you. Your solution helped me to write a full object oriented abstraction on a JavaScript library (Fabric JS) completely in C#. The finding with the reviver is really great. I just hope MS is not going to remove it :)

**Comment by TFTomSun** [comment]
I've added another fix. The call of getInstanceMethod always returned the return value of the invoked method. But that's unintended when I call it as a action without expecting any return value. In my concrete case the invoked method of an API returned a object and the signalr message size was exceeded. I was wondering why that happened, because I just called it as an action... I didn't even know that the API method has a return value.

File: src/BrowserInterop/BrowserInterop.csproj [hunk 1]
```diff
@@ -6,7 +6,7 @@
     <PackageId>BrowserInterop</PackageId>
     <Authors>Rémi BOURGAREL</Authors>
     <Company>Rémi BOURGAREL</Company>
-    <PackageVersion>1.1.1</PackageVersion>
+    <PackageVersion>1.1.1.13</PackageVersion>
```
Comment (RemiBou) [inline src/BrowserInterop/BrowserInterop.csproj:9]: You can remove this I'll do a patch release I think (1.1.2) with an other bug fix, then a minor upgrade for .net 5 (1.2)

File: src/BrowserInterop/Extensions/BrowserInteropJsRuntimeExtension.cs [hunk 2]
```diff
@@ -151,7 +151,7 @@ public static async ValueTask<WindowInterop> Window(this IJSRuntime jsRuntime)
         public static async ValueTask InvokeInstanceMethod(this IJSRuntime jsRuntime, JsRuntimeObjectRef windowObject,
             string methodName, params object[] arguments)
         {
-            await jsRuntime.InvokeVoidAsync("browserInterop.callInstanceMethod",
+            await jsRuntime.InvokeVoidAsync("browserInterop.callInstanceAction",
```
Comment (RemiBou) [inline src/BrowserInterop/Extensions/BrowserInteropJsRuntimeExtension.cs:154]: what's the point of this renaming?

File: src/BrowserInterop/Extensions/BrowserInteropJsRuntimeExtension.cs [hunk 3]
```diff
@@ -225,11 +226,36 @@ public static async ValueTask<WindowInterop> Window(this IJSRuntime jsRuntime)
...
+        /// </summary>
+        /// <param name="jsRuntime">Current JS Runtime</param>
+        /// <param name="windowObject">Reference to the JS instance</param>
+        /// <param name="methodName">Method name/path </param>
+        /// <param name="arguments">method arguments</param>
+        /// <returns></returns>
+        public static async ValueTask<JsRuntimeObjectRef[]> InvokeInstanceMethodGetRefs(this IJSRuntime jsRuntime,
+            JsRuntimeObjectRef windowObject, string methodName, params object[] arguments)
+        {
+            if (jsRuntime is null) throw new ArgumentNullException(nameof(jsRuntime));
```
Comment (RemiBou) [inline src/BrowserInterop/Extensions/BrowserInteropJsRuntimeExtension.cs:246]: nit : I'm not sure this is relevant, better be safe than sorry anyway

File: src/BrowserInterop/Extensions/CallBackInteropWrapper.cs [hunk 4]
```diff
@@ -18,9 +18,34 @@ public class CallBackInteropWrapper
...
+        /// or if it should be postponed until no event is incomming within one debounce time period.
+        /// </summary>
+        public bool TriggerPermanent { get; set; }
+
+        /// <summary>
+        /// Determines whether default values of the serialized callback data shall be transmitted
+        /// </summary>
+        public bool IncludeDefaults { get; set; }
+
+        public bool GetArgumentsSerializationAndRef { get; set; }
```
Comment (RemiBou) [inline src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:44]: While I understand the use case, I'm not a big fan of this being only here. Maybe we could use the serialization spec for saying we want to get a reference to the js variable (instead of "*" use "&").

File: src/BrowserInterop/Extensions/CallBackInteropWrapper.cs [hunk 5]
```diff
@@ -86,15 +114,31 @@ private CallBackInteropWrapper()
...
+            ((IDisposable)this.CallbackRef).Dispose();
+        }
+
+
+        public object CallbackRef { get;  } // we must keep it as a object, otherwise its not serialized.
+    }
+    internal interface ICallbackReferenceData
+    {
+    }
+    public class CallbackReferenceData<T>: ICallbackReferenceData
```
Comment (RemiBou) [inline src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:138]: Like in previous comment, this could be pulled up and use in other js interop utilities where I do get ref then get data.

File: src/BrowserInterop/Extensions/CallBackInteropWrapper.cs [hunk 6]
```diff
@@ -86,15 +114,31 @@ private CallBackInteropWrapper()
...
+        public void Dispose()
+        {
+            //casting is safe, the constructor enforces a disposable.
+            ((IDisposable)this.CallbackRef).Dispose();
+        }
+
+
+        public object CallbackRef { get;  } // we must keep it as a object, otherwise its not serialized.
+    }
+    internal interface ICallbackReferenceData
```
Comment (RemiBou) [inline src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:135]: What's the point of this interface ?

File: src/BrowserInterop/Extensions/CallBackInteropWrapper.cs [hunk 7]
```diff
@@ -86,15 +114,31 @@ private CallBackInteropWrapper()
...
             {
-                CallbackRef = DotNetObjectReference.Create(new JsInteropActionWrapper(callback)),
                 SerializationSpec = serializationSpec,
                 GetJsObjectRef = getJsObjectRef
             };
             return res;
         }
 
-        public object CallbackRef { get; set; }
+        public void Dispose()
```
Comment (RemiBou) [inline src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:126]: is it because there was a leak and the callbackref was never cleaned up?

File: src/BrowserInterop/wwwroot/scripts.js [hunk 8]
```diff
@@ -138,6 +191,9 @@ browserInterop = new (function () {
         var res = me.getSerializableObject(data, [], serializationSpec);
         return res;
     };
+    this.callInstanceAction = function (instance, methodPath, ...args) {
```
Comment (RemiBou) [inline src/BrowserInterop/wwwroot/scripts.js:194]: all the other methods use "Method", I don't think this renaming is relevant here

File: src/BrowserInterop/Extensions/CallBackInteropWrapper.cs [hunk 9]
```diff
@@ -18,9 +18,34 @@ public class CallBackInteropWrapper
...
+        /// <summary>
+        /// Determines whether the callback shall be invoked always when the debounce time is expired
+        /// or if it should be postponed until no event is incomming within one debounce time period.
+        /// </summary>
+        public bool TriggerPermanent { get; set; }
+
+        /// <summary>
+        /// Determines whether default values of the serialized callback data shall be transmitted
+        /// </summary>
+        public bool IncludeDefaults { get; set; }
```
Comment (RemiBou) [inline src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:42]: what is the point of this instead of passing "*"?

**Review by RemiBou — CHANGES_REQUESTED** [review]
Thanks for this PR again. While I agree with most of your change, I need more coverage in testing. Testing is done via e2e tests only, I don't need 100% coverage but a few tests with some edge case you can think about would reassure me.

File: src/BrowserInterop/Extensions/BrowserInteropJsRuntimeExtension.cs (hunk 2 above)
Comment (TFTomSun) [reply src/BrowserInterop/Extensions/BrowserInteropJsRuntimeExtension.cs:154]: callInstanceAction is a new method, not a rename. 'Action' indicates that there is no return value, like in .NET the Func / Action delegates. I woud prefer to follow the .NET namings: Function for methods with return value and Action for methods without return value

Comment (TFTomSun) [reply src/BrowserInterop/Extensions/BrowserInteropJsRuntimeExtension.cs:154]: The problem with browserInterop.callInstanceMethod is that it always tries to serialize the return value of the invoke method. But I as a user want to decide whether I need the return value or not, and I do that already by calling ValueTask InvokeInstanceMethod which indicates that I don't expect / don't need the return value. That shall be respected by the javascript side.

File: src/BrowserInterop/Extensions/BrowserInteropJsRuntimeExtension.cs (hunk 3 above)
Comment (TFTomSun) [reply src/BrowserInterop/Extensions/BrowserInteropJsRuntimeExtension.cs:246]: what do you mean? I had a scenario where a javascript method returns an array of objects that I want to treat as references on the .NET side. Therefore I added that method. Is there another way to do that already?

File: src/BrowserInterop/BrowserInterop.csproj (hunk 1 above)
Comment (TFTomSun) [reply src/BrowserInterop/BrowserInterop.csproj:9]: I just need that versioning for local pack and wildcard based restore, to get automatically the latest version I've built locally. I used the 4th digit to automatically 1.1.2 once it is released. feel free to pull the version up to 1.1.2 when you do a release.

File: src/BrowserInterop/Extensions/CallBackInteropWrapper.cs (hunk 4 above)
Comment (TFTomSun) [reply src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:44]: The main use case of callbacks are events I think. So mostly you might directly want the event args deserialized. But some event args also provide functions (e.g. mouse event args  => args.preventDefault()) . I think its better to provide the serialized value by default but optionally get the reference as well. I want to reduce the amount of interop calls. If I register for mouse events I don't want to call the objects data whenever the event occurs. On the other hand I might want to call the preventDefault() method, therefore I need a possibility to get both: The event data and the args instance.

File: src/BrowserInterop/Extensions/CallBackInteropWrapper.cs (hunk 5 above)
Comment (TFTomSun) [reply src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:138]: Some events like mouse move events occur really often, as described above I would prefer to avoid the call back from .NET to javascript on every event, just to get the events data.

File: src/BrowserInterop/Extensions/CallBackInteropWrapper.cs (hunk 7 above)
Comment (TFTomSun) [reply src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:126]: yes it wasn't made visible that there's a disposable resource under that wrapper (the DotnetCallback) . By adding the IDisposable I just made it visible so that the user is aware of it.

File: src/BrowserInterop/wwwroot/scripts.js (hunk 8 above)
Comment (TFTomSun) [reply src/BrowserInterop/wwwroot/scripts.js:194]: but then you would have a conflict with the existing callInstanceMethod

File: src/BrowserInterop/Extensions/CallBackInteropWrapper.cs (hunk 9 above)
Comment (TFTomSun) [reply src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:42]: lets say we have a classic person with name and address datastructure. Your current approach was to use the anonymous types only as serialization filters (e.g.):

`runtime.GetInstanceProperty<MyPerson>(objectRef,"person",new { FirstName = "*", Age="*", Address="*"})`

but I use the serialization filter as the data structure type itself

var data = await runtime.MyOwnGetInstanceProperty(objectRef,"person",new { FirstName = default(string) Age=default(int)})
Asssert.AreEqual(30,data.Age);

Therefore the string based serialization filtering is not suitable for me. I need a filter that is applied just by the anonymous properties and not by their value.

File: src/BrowserInterop/Extensions/CallBackInteropWrapper.cs (hunk 6 above)
Comment (TFTomSun) [reply src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:135]: I use it to determine whether the user requests the data and the ref. see CallbackInteropWrapper line 96

**Comment by TFTomSun** [comment]
regarding tests: I can't load the test project, I don't know what type of project that is. If it would be a standard csharp project (e.g. mstest + selenium) I would have added some tests.

File: src/BrowserInterop/Extensions/BrowserInteropJsRuntimeExtension.cs (hunk 2 above)
Comment (RemiBou) [reply src/BrowserInterop/Extensions/BrowserInteropJsRuntimeExtension.cs:154]: Ok I see cool

File: src/BrowserInterop/Extensions/BrowserInteropJsRuntimeExtension.cs (hunk 3 above)
Comment (RemiBou) [reply src/BrowserInterop/Extensions/BrowserInteropJsRuntimeExtension.cs:246]: No I meant the null check, the new method is great :)

File: src/BrowserInterop/Extensions/CallBackInteropWrapper.cs (hunk 9 above)
Comment (RemiBou) [reply src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:42]:
```cs
runtime.GetInstanceProperty<MyPerson>(objectRef,"person","*")
```

will work too. The thing is, you can use "*" nearly everywhere. The serialization spec are here mostly for browser api where there can be loops or a lot of data (like window.document) that we don't want/need on .net side

File: src/BrowserInterop/Extensions/CallBackInteropWrapper.cs (hunk 4 above)
Comment (RemiBou) [reply src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:44]: I wasn't clear sorry: I totally agree with you, this is nice to avoid 2 interop call. What I would like is to use this mechanismon method call / property get as well.

File: src/BrowserInterop/Extensions/CallBackInteropWrapper.cs (hunk 6 above)
Comment (RemiBou) [reply src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:135]: why not using the class CallbackReferenceData directly, here we leave the door open for implementation of ICallbackReferenceData without Ref/Data fields

File: src/BrowserInterop/Extensions/CallBackInteropWrapper.cs (hunk 5 above)
Comment (RemiBou) [reply src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:138]: we'll keep the conversation on the other comment

File: src/BrowserInterop/wwwroot/scripts.js (hunk 8 above)
Comment (RemiBou) [reply src/BrowserInterop/wwwroot/scripts.js:194]: after reading your reason for this method we can keep it like that

**Comment by RemiBou** [comment]
I added some instruction for testing to README, it's a cypress project. You need to install npm, then browse to the E2E tests directory and do "npm install && npm run-script start"

File: src/BrowserInterop/Extensions/CallBackInteropWrapper.cs (hunk 9 above)
Comment (TFTomSun) [reply src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:42]: have a closer look at my sample. I don't specify any generic parameter. The type safetyness comes from the anonymous type. That can safe alot of boilerplate code on .NET side. I even considered to add some naming convention based access of sub object properties like: 
var data = await runtime.MyOwnGetInstanceProperty(objectRef,"person",new { Address__StreetNumber= default(int))
Asssert.AreEqual(10,data.Address__StreetNumber);

where the anonymous type property Address__StreetNumber is automatically translated to x['address']['streetNumber']

Comment (TFTomSun) [reply src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:42]: What I already implemented too is typesafe filtering using expression trees. Means: I have a data structure, but i am able to filter in a typesafe manner for some of its properties e.g.:

```
var anonymousData = runtime.GetInstanceProperty<MyPerson>(objectRef,"person",x=> new {x.Name, x.Address.StreeNumber});
Asssert.AreEqual(10, anonyousData.StreetNumber);
```

Comment (TFTomSun) [reply src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:42]: additional note: I just realized the whole processing of my examples could be done and cached on .NET side, by creating a Dictionary or JsonDocument serialization filter object per anonymous type and passing these filter objects then to your API. That means:  The change in js script would not be necessary. I'll consider that approach, because it might also reduce the serialization effort on .NET side if  you call a method with the same filter multiple times.

File: src/BrowserInterop/Extensions/CallBackInteropWrapper.cs (hunk 6 above)
Comment (TFTomSun) [reply src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:135]: because CallbackReferenceData is a generic type and the type checking would be more heavy. The ICallbackReferenceData interface is internal. At least the API user cannot implement it. This concept is called marker interface.

File: src/BrowserInterop/Extensions/CallBackInteropWrapper.cs (hunk 4 above)
Comment (TFTomSun) [reply src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:44]: sure that can be done. I didn't have the requirement yet.

File: src/BrowserInterop/Extensions/BrowserInteropJsRuntimeExtension.cs (hunk 3 above)
Comment (TFTomSun) [reply src/BrowserInterop/Extensions/BrowserInteropJsRuntimeExtension.cs:246]: I can remove the null check. Thats ok for me :)

File: src/BrowserInterop/Extensions/CallBackInteropWrapper.cs (hunk 4 above)
Comment (TFTomSun) [reply src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:44]: If I pull all the typesafe filter handling up to the client side, I could also easily create those filters you mentioned ("&" for reference "*" for serialization value)

Comment (TFTomSun) [reply src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:44]: what do you think about using "&[]" for indicating that you want the references of array items instead of the array reference? Do you have other use cases that might it be worth to implement?

Comment (TFTomSun) [reply src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:44]: Is it already possible today to exclude properties? For example Person =>  I want to include the whole address except the streetNumber

Comment (RemiBou) [reply src/BrowserInterop/Extensions/CallBackInteropWrapper.cs:44]: "[&]" is ok, I don't mind adding more syntax to serializationSpec. We can also add person:{"*":true,"address":"!"} what do you think ?

**Comment by RemiBou** [comment]
Sorry closing, not sure I actually want this feature, please open a suggestion ticket before proposing important changes
